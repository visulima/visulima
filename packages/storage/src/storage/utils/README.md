# Storage Utilities

Shared utility functions and classes used across all storage backends in the Visulima upload system. These utilities provide common functionality for file handling, metadata management, and storage operations.

## Directory Structure

```
utils/
├── file/                    # File-related utilities
│   ├── index.ts            # Main exports
│   ├── file.ts             # Core file types and classes
│   ├── metadata.ts         # Metadata handling
│   ├── extract-mime-type.ts # MIME type extraction
│   ├── extract-original-name.ts # Filename extraction
│   ├── file-name.ts        # File naming utilities
│   ├── get-file-status.ts  # Upload status calculation
│   ├── has-content.ts      # Content checking
│   ├── is-expired.ts       # Expiration checking
│   ├── part-match.ts       # Upload part matching
│   ├── types.ts            # Type definitions
│   └── update-metadata.ts  # Metadata updates
└── README.md               # This file
```

## Core File Types

### UploadFile Interface

The core file interface used across all storage backends:

```typescript
interface UploadFile {
    bytesWritten: number; // Bytes uploaded so far
    contentType: string; // MIME type
    createdAt?: Date | number; // Creation timestamp
    expiredAt?: Date | number; // Expiration timestamp
    hash?: FileHash; // File hash information
    id: string; // Unique file identifier
    metadata?: Record<string, any>; // Custom metadata
    modifiedAt?: Date | number; // Last modified timestamp
    name?: string; // Original filename
    size?: number; // File size in bytes
    status?: UploadEventType; // Upload status
}
```

### File Classes

#### S3File, GCSFile, AzureFile

Provider-specific file implementations:

```typescript
class S3File extends UploadFile {
    // AWS S3 specific properties and methods
    bucket: string;

    key: string;

    etag?: string;
}

class GCSFile extends UploadFile {
    // Google Cloud Storage specific properties
    bucket: string;

    name: string;

    generation?: string;
}

class AzureFile extends UploadFile {
    // Azure Blob Storage specific properties
    container: string;

    blobName: string;

    etag?: string;
}
```

## File Operations

### File Status Management

```typescript
import { getFileStatus } from "./utils/file/get-file-status";

// Determine upload status based on bytes written vs total size
const status = getFileStatus(file);
// Returns: 'created', 'part', 'completed'
```

### Content Validation

```typescript
import { hasContent } from "./utils/file/has-content";

// Check if upload part has actual content
if (hasContent(part)) {
    // Process the content
}
```

### Expiration Checking

```typescript
import { isExpired } from "./utils/file/is-expired";

// Check if file has expired
if (isExpired(file)) {
    // Handle expired file
}
```

## Metadata Management

### Metadata Handling

```typescript
import { Metadata } from "./utils/file/metadata";

// Create metadata object
const metadata = new Metadata({
    contentType: "image/jpeg",
    filename: "photo.jpg",
});

// Serialize for storage
const serialized = metadata.toString();

// Parse from storage
const parsed = Metadata.parse(serializedMetadata);
```

### Metadata Updates

```typescript
import { updateMetadata } from "./utils/file/update-metadata";

// Update file metadata
const updatedFile = updateMetadata(file, {
    processed: true,
    thumbnailUrl: "/thumbnails/photo.jpg",
});
```

## MIME Type Handling

### MIME Type Extraction

```typescript
import { extractMIMEType } from "./utils/file/extract-mime-type";

// Extract MIME type from request
const mimeType = extractMIMEType(request);
// Returns: 'image/jpeg', 'application/pdf', etc.
```

### Filename Extraction

```typescript
import { extractOriginalName } from "./utils/file/extract-original-name";

// Extract original filename from request
const filename = extractOriginalName(request);
// Returns: 'photo.jpg', 'document.pdf', etc.
```

## File Naming

### File Name Generation

```typescript
import { FileName } from "./utils/file/file-name";

// Generate unique file names
const filename = FileName.generate("photo.jpg");
// Returns: 'photo-abc123.jpg'

// Validate filenames
if (FileName.isValid(filename)) {
    // Filename is safe to use
}

// Configure naming patterns
FileName.configure({
    prefix: "upload-",
    suffix: "-processed",
});
```

## Upload Part Management

### Part Matching

```typescript
import { partMatch } from "./utils/file/part-match";

// Check if upload part matches expected chunk
if (partMatch(part, expectedStart, expectedSize)) {
    // Part is valid
}
```

### Size Updates

```typescript
import { updateSize } from "./utils/file/update-size";

// Update file size information
const updatedFile = updateSize(file, newBytesWritten);
```

## Utility Functions

### Checksum Validation

```typescript
// Checksum validation is handled by individual storage backends
// using the checksumTypes configuration

const storage = new S3Storage({
    bucket: "my-bucket",
    genericConfig: {
        checksumTypes: ["md5", "crc32c", "sha256"],
    },
});
```

### Hash Generation

File hashing for integrity verification:

```typescript
// Hash information stored in file metadata
interface FileHash {
    algorithm: string; // 'md5', 'sha1', 'sha256', etc.
    value: string; // Hex-encoded hash value
}
```

## Error Handling

### File Operation Errors

```typescript
import { ERRORS } from "../utils";

// Common error codes
ERRORS.FILE_ERROR; // General file operation error
ERRORS.FILE_NOT_FOUND; // File not found
ERRORS.INVALID_CHECKSUM; // Checksum validation failed
```

## Testing Utilities

### Mock Files

Create mock files for testing:

```typescript
import { File } from "./utils/file/file";

// Create test file
const testFile = new File({
    contentType: "image/jpeg",
    id: "test-123",
    metadata: { test: true },
    size: 1_024_000,
});
```

## Performance Considerations

### Memory Management

```typescript
// Large file handling
const storage = new DiskStorage({
    directory: "/uploads",
    // Configure buffer sizes for optimal memory usage
    highWaterMark: 64 * 1024, // 64KB chunks
});
```

### Concurrent Access

```typescript
// File locking prevents corruption
await storage.write({
    body: data,
    contentLength: data.length,
    id: file.id,
    start: 0,
});
// Automatic locking during operations
```

## Configuration

### Utility Configuration

```typescript
// Configure global utility behavior
FileName.configure({
    invalidPrefixes: ["..", "."],
    invalidSuffixes: [".exe", ".bat"],
    maxLength: 255,
});
```

## Development Guidelines

### Adding New Utilities

1. **File Organization**: Place utilities in appropriate subdirectories
2. **Type Safety**: Ensure all utilities are fully typed
3. **Error Handling**: Provide clear error messages
4. **Testing**: Include comprehensive test coverage
5. **Documentation**: Document all public APIs

### Utility Design Principles

- **Pure Functions**: Prefer pure functions over classes where possible
- **Immutability**: Avoid mutating input objects
- **Type Safety**: Leverage TypeScript for compile-time safety
- **Performance**: Optimize for common use cases
- **Compatibility**: Work across all storage backends

## Common Patterns

### File Processing Pipeline

```typescript
// Typical file processing flow
const processFile = async (request: Request) => {
    // 1. Extract metadata
    const mimeType = extractMIMEType(request);
    const filename = extractOriginalName(request);

    // 2. Create file record
    const file = await storage.create(request, {
        contentType: mimeType,
        metadata: { source: "upload" },
        originalName: filename,
    });

    // 3. Process upload
    // ... upload logic ...

    // 4. Update metadata
    const processedFile = updateMetadata(file, {
        processed: true,
        processedAt: new Date(),
    });

    return processedFile;
};
```

### Error Recovery

```typescript
// Handle upload interruptions
const resumeUpload = async (fileId: string) => {
    const file = await storage.get({ id: fileId });

    if (isExpired(file)) {
        throw new Error("Upload expired");
    }

    if (getFileStatus(file) === "completed") {
        return file; // Already done
    }

    // Resume from last position
    return file;
};
```

## Troubleshooting

### Common Issues

1. **Invalid File Names**

    ```typescript
    // Check filename validity
    if (!FileName.isValid(filename)) {
        throw new Error("Invalid filename");
    }
    ```

2. **Memory Issues with Large Files**

    ```typescript
    // Use streaming for large files
    const stream = fs.createReadStream(largeFile);

    await storage.write({
        body: stream,
        contentLength: stats.size,
        id: file.id,
        start: 0,
    });
    ```

3. **Metadata Corruption**

    ```typescript
    // Validate metadata before saving
    const validMetadata = sanitizeMetadata(rawMetadata);

    await storage.update({ id: file.id }, { metadata: validMetadata });
    ```

## Contributing

When adding new utilities:

1. **Test Coverage**: Ensure 100% test coverage
2. **Type Definitions**: Provide complete TypeScript types
3. **Documentation**: Update this README with new utilities
4. **Backwards Compatibility**: Don't break existing APIs
5. **Performance**: Profile and optimize for performance
