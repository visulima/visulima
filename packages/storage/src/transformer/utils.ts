import mime from "mime/lite";

import type { FileReturn } from "../storage/utils/file";

/**
 * Check if a content type is valid for a specific media type
 * @param contentType MIME content type string
 * @param expectedType Expected media type ('image', 'video', or 'audio')
 * @returns True if the content type is valid for the expected media type
 */
export const isValidMediaType = (contentType: string | undefined, expectedType: "image" | "video" | "audio"): boolean => {
    if (!contentType) {
        return false;
    }

    return contentType.startsWith(`${expectedType}/`);
};

/**
 * Get format (extension) from content type using mime package
 * @param contentType MIME content type string
 * @returns Format string or undefined if not found
 */
export const getFormatFromContentType = (contentType: string | undefined): string | undefined => {
    if (!contentType) {
        return undefined;
    }

    return mime.getExtension(contentType) || undefined;
};

/**
 * Validate a media file for a specific type
 * @param file File to validate
 * @param expectedType Expected media type
 * @param config Validation configuration
 * @throws Error if validation fails
 */
export const validateMediaFile = (
    file: FileReturn,
    expectedType: "image" | "video" | "audio",
    config?: {
        maxSize?: number;
        supportedFormats?: string[];
    },
): void => {
    // Check file size
    const fileSize = typeof file.size === "string" ? Number.parseInt(file.size, 10) : file.size;

    if (config?.maxSize && fileSize > config.maxSize) {
        throw new Error(`${expectedType} size ${fileSize} exceeds maximum allowed size ${config.maxSize}`);
    }

    // Check if it's the expected media type
    if (!isValidMediaType(file.contentType, expectedType)) {
        throw new Error(`File is not ${expectedType === "image" ? "an" : "a"} ${expectedType}: ${file.contentType}`);
    }

    // Check format support
    const format = getFormatFromContentType(file.contentType);

    if (config?.supportedFormats && format && !config.supportedFormats.includes(format)) {
        throw new Error(`Unsupported ${expectedType} format: ${format}`);
    }
};

/**
 * Validate that a content type is known and supported by the mime package
 * @param contentType MIME content type string
 * @returns True if the content type is known
 */
export const isKnownContentType = (contentType: string | undefined): boolean => {
    if (!contentType) {
        return false;
    }

    return !!mime.getExtension(contentType);
};
