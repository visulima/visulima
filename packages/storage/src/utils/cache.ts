/* eslint-disable class-methods-use-this */

/**
 * Simple cache interface that any cache implementation can follow
 */
export interface Cache<K = string, V = any> {
    /** Clear all cache entries */
    clear: () => void | Promise<void>;

    /** Delete a value from cache */
    delete: (key: K) => boolean | Promise<boolean>;

    /** Get a value from cache */
    get: (key: K) => V | undefined | Promise<V | undefined>;

    /** Check if a key exists in cache */
    has: (key: K) => boolean | Promise<boolean>;

    /** Set a value in cache */
    set: (key: K, value: V, options?: CacheOptions) => boolean | Promise<boolean>;
}

/**
 * Options for cache operations
 */
export interface CacheOptions {
    /** TTL in milliseconds */
    ttl?: number;
}

/**
 * No-op cache implementation that does nothing (used when caching is disabled)
 */
export class NoOpCache<K = string, V = any> implements Cache<K, V> {
    public get(): undefined {
        return undefined;
    }

    public set(): boolean {
        return true;
    }

    public delete(): boolean {
        return true;
    }

    public clear(): void {
        // Do nothing
    }

    public has(): boolean {
        return false;
    }
}
