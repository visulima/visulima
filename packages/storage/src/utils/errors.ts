import mem from "memoize";

import type { HttpError } from "./types";

/**
 * Canonical error codes used across handlers and storage adapters.
 * These codes map to standardized HTTP status codes and error messages.
 */
export enum ERRORS {
    BAD_REQUEST = "BadRequest",
    CHECKSUM_MISMATCH = "ChecksumMismatch",
    FILE_CONFLICT = "FileConflict",
    FILE_ERROR = "FileError",
    FILE_LOCKED = "FileLocked",
    FILE_NOT_ALLOWED = "FileNotAllowed",
    FILE_NOT_FOUND = "FileNotFound",
    FORBIDDEN = "Forbidden",
    GONE = "Gone",
    INVALID_CONTENT_TYPE = "InvalidContentType",
    INVALID_FILE_NAME = "InvalidFileName",
    INVALID_FILE_SIZE = "InvalidFileSize",
    INVALID_RANGE = "InvalidRange",
    METHOD_NOT_ALLOWED = "MethodNotAllowed",
    REQUEST_ABORTED = "RequestAborted",
    REQUEST_ENTITY_TOO_LARGE = "RequestEntityTooLarge",
    STORAGE_BUSY = "StorageBusy",
    STORAGE_ERROR = "StorageError",
    TOO_MANY_REQUESTS = "TooManyRequests",
    UNKNOWN_ERROR = "UnknownError",
    UNPROCESSABLE_ENTITY = "UnprocessableEntity",
    // eslint-disable-next-line no-secrets/no-secrets
    UNSUPPORTED_CHECKSUM_ALGORITHM = "UnsupportedChecksumAlgorithm",
    UNSUPPORTED_MEDIA_TYPE = "UnsupportedMediaType",
}

/**
 * Type mapping of error codes to standardized HTTP error responses.
 * @template T - The error code type (defaults to string)
 */
export type ErrorResponses<T extends string = string> = {
    [K in T]: HttpError;
};

/**
 * Lazily materialized mapping of error codes to HttpError response objects.
 * Uses memoization to avoid recomputing the error map on each access.
 * @returns Map of error codes to standardized HTTP error responses
 */
export const ErrorMap = mem((): ErrorResponses => {
    const errors: Record<string, [number, string]> = {
        BadRequest: [400, "Bad request"],
        ChecksumMismatch: [460, "Checksum mismatch"],
        FileConflict: [409, "File conflict"],
        FileError: [500, "Something went wrong writing the file"],
        FileLocked: [423, "File locked"],
        FileNotAllowed: [403, "File not allowed"],
        FileNotFound: [404, "Not found"],
        Forbidden: [403, "Authenticated user is not allowed access"],
        Gone: [410, "The file for this url no longer exists"],
        InvalidFileName: [400, "Invalid file name or it cannot be retrieved"],
        InvalidFileSize: [400, "File size cannot be retrieved"],
        InvalidRange: [400, "Invalid or missing content-range header"],
        Invalidtype: [400, "Invalid or missing \"content-type\" header"],
        MethodNotAllowed: [405, "Method not allowed"],
        RequestAborted: [499, "Request aborted"],
        RequestEntityTooLarge: [413, "Request entity too large"],
        StorageBusy: [503, "Storage is busy"],
        StorageError: [503, "Storage error"],
        TooManyRequests: [429, "Too many requests"],
        UnknownError: [500, "Something went wrong"],
        UnprocessableEntity: [422, "Validation failed"],
        UnsupportedChecksumAlgorithm: [400, "Unsupported checksum algorithm"],
        UnsupportedMediaType: [415, "Unsupported media type"],
    };

    const errorMap: ErrorResponses = {};

    (Object.keys(errors) as ERRORS[]).forEach((code) => {
        const [statusCode, message] = errors[code] as [number, string];

        errorMap[code] = { code, message, statusCode };
    });

    return errorMap;
})();

/**
 * Error subclass carrying a stable error code and optional detail.
 * Provides structured error information for upload operations.
 */
export class UploadError extends Error {
    public override name = "UploadError";

    /** The standardized error code from the ERRORS enum */
    public UploadErrorCode: ERRORS = ERRORS.UNKNOWN_ERROR;

    /** Optional additional error details */
    public detail?: unknown;

    /**
     * Creates a new UploadError instance.
     * @param code Standardized error code (defaults to UNKNOWN_ERROR)
     * @param message Human-readable error message (defaults to the code)
     * @param detail Optional additional error details
     */
    public constructor(code: ERRORS = ERRORS.UNKNOWN_ERROR, message?: string, detail?: unknown) {
        super(message || code);
        this.name = "UploadError";
        this.detail = detail;

        if (Object.values(ERRORS).includes(code)) {
            this.UploadErrorCode = code;
        }
    }
}

/**
 * Type guard to check if an error is an UploadError instance.
 * @param error Error to check
 * @returns True if the error is an UploadError with a valid error code
 */
export const isUploadError = (error: unknown): error is UploadError => !!(error as UploadError).UploadErrorCode;

/**
 * Convenience function to throw an UploadError from a string error code.
 * Looks up the appropriate error message from ErrorMap.
 * @param UploadErrorCode String error code to convert to UploadError
 * @param detail Optional additional error details
 * @throws UploadError with the specified code and message
 */
export const throwErrorCode = (UploadErrorCode: string, detail?: string): never => {
    const error = new UploadError(detail || (ErrorMap[UploadErrorCode] as HttpError).message);

    error.UploadErrorCode = UploadErrorCode as ERRORS;

    if (typeof detail === "string") {
        error.detail = detail;
    }

    throw error;
};
