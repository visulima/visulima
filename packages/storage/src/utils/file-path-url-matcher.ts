import { match } from "path-to-regexp";

const filePathUrlMatcher = match<PathMatch>(["/*path/:uuid/:metadata", "/*path/:uuid.:ext", "/*path/:uuid"], {
    decode: decodeURIComponent,
});

export type PathMatch = { ext?: string; metadata?: string; uuid: string };

export default filePathUrlMatcher;
