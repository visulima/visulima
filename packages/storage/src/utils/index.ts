import fnv1a from "@sindresorhus/fnv1a";
import mem from "memoize";

export const hash = mem((value: string) => fnv1a(value, { size: 64 }).toString(16));

export type { ErrorResponses } from "./errors";
export { ErrorMap, ERRORS, isUploadError, throwErrorCode, UploadError } from "./errors";
export type { PathMatch } from "./file-path-url-matcher";
export { default as filePathUrlMatcher } from "./file-path-url-matcher";
export {
    appendHeader,
    extractHost,
    extractProto,
    getBaseUrl,
    getHeader,
    getIdFromRequest,
    getMetadata,
    getRealPath,
    getRequestStream,
    normalizeHookResponse,
    normalizeOnErrorResponse,
    readBody,
    setHeaders,
    uuidRegex,
} from "./http";
export { default as Locker } from "./locker";
export { StreamChecksum, streamChecksum } from "./pipes/stream-checksum";
export { default as StreamLength } from "./pipes/stream-length";
export { default as isEqual } from "./primitives/is-equal";
export { default as isRecord } from "./primitives/is-record";
export { default as mapValues } from "./primitives/map-values";
export { default as pick } from "./primitives/pick";
export { default as toMilliseconds } from "./primitives/to-milliseconds";
export { default as toSeconds } from "./primitives/to-seconds";
export { default as RangeHasher } from "./range-hasher";
export type {
    Header,
    Headers,
    HttpError,
    HttpErrorBody,
    IncomingMessageWithBody,
    Logger,
    ResponseBody,
    ResponseBodyType,
    ResponseTuple,
    UploadResponse,
    Validation,
    ValidationError,
    ValidatorConfig,
} from "./types";
export { isValidationError, Validator } from "./validator";
