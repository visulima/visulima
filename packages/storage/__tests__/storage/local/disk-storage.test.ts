import { join } from "node:path";

import { fs, vol } from "memfs";
import { createRequest } from "node-mocks-http";
import { temporaryDirectory } from "tempy";
import { afterAll, beforeAll, beforeEach, describe, expect, expectTypeOf, it, vi } from "vitest";

import DiskStorage from "../../../src/storage/local/disk-storage";
import { metafile, storageOptions } from "../../__helpers__/config";
import FileWriteStream from "../../__helpers__/streams/file-write-stream";
import RequestReadStream from "../../__helpers__/streams/request-read-stream";
import { deepClone } from "../../__helpers__/utils";

// Mock fs modules with memfs
vi.mock(import("node:fs"), async () => {
    const actual = await vi.importActual("node:fs");

    return {
        ...actual,
        ...fs,
    };
});
vi.mock(import("node:fs"), async () => {
    const actual = await vi.importActual("node:fs");

    return {
        ...actual,
        ...fs,
    };
});
vi.mock(import("node:fs/promises"), async () => {
    const actual = await vi.importActual("node:fs/promises");

    return {
        ...actual,
        ...fs.promises,
    };
});
vi.mock(import("@visulima/fs"), async () => {
    const actual = await vi.importActual("@visulima/fs");

    return {
        ...actual,
        ensureFile: async (path: string) => {
            // Ensure directory exists
            const dir = path.slice(0, Math.max(0, path.lastIndexOf("/")));

            await fs.promises.mkdir(dir, { recursive: true });
            // Create empty file
            await fs.promises.writeFile(path, "");
        },
        remove: fs.promises.rm,
        walk: actual.walk, // Keep original walk if it doesn't use fs
    };
});

describe(DiskStorage, () => {
    let directory: string;

    vi.useFakeTimers();
    vi.setSystemTime(new Date("2022-02-02"));

    let options: typeof storageOptions;
    let storage: DiskStorage;
    let readStream: RequestReadStream;

    const request = createRequest();

    const createFile = async (createDummyFile: boolean = true): Promise<void> => {
        storage = new DiskStorage(options);

        const filePath = join(directory, "testfile.mp4");

        await (createDummyFile ? fs.promises.writeFile(filePath, "") : fs.promises.rm(filePath));

        await storage.create(request, deepClone(metafile));
    };

    beforeAll(async () => {
        directory = temporaryDirectory();
        options = { ...storageOptions, directory };
        // Set up memfs volume
        vol.fromJSON({}, directory);
    });

    afterAll(async () => {
        try {
            const { rm } = await import("node:fs/promises");

            await rm(directory, { force: true, recursive: true });
        } catch {
            // ignore if directory doesn't exist
        }
    });

    describe("initialization", () => {
        it("should set directory path correctly", () => {
            expect.assertions(1);

            storage = new DiskStorage(options);

            expect(storage.directory).toBe(directory);
        });
    });

    describe(".create()", () => {
        beforeEach(() => {
            storage = new DiskStorage(options);
        });

        it("should create disk file with correct status and bytesWritten", async () => {
            expect.assertions(4);

            const diskFile = await storage.create(request, metafile);

            expect(diskFile.id).toBeDefined();
            expect(diskFile.bytesWritten).toBe(0);
            expect(diskFile.contentType).toBe(metafile.contentType);
            expect(diskFile.status).toBe("created");

            expectTypeOf(diskFile.expiredAt).toBeNumber();
        });

        it("should reject creation when file size exceeds limits", async () => {
            expect.assertions(2);

            await expect(storage.create(request, { ...metafile, size: 6e10 })).rejects.toThrow();
            await expect(storage.create(request, { ...metafile, size: 6e10 })).rejects.toThrow(/Request entity too large|ValidationError/);
        });
    });

    describe(".update()", () => {
        beforeEach(createFile);

        it("should update file metadata correctly", async () => {
            expect.assertions(2);

            const file = await storage.update(metafile, { metadata: { name: "newname.mp4" } });

            expect(file.metadata.name).toBe("newname.mp4");
            expect(file.metadata.mimeType).toBe("video/mp4");
        });
    });

    describe(".write()", () => {
        beforeEach(() => {
            readStream = new RequestReadStream();

            createFile();
        });

        it.todo("should write data and update file status and bytesWritten", async () => {
            expect.assertions(2);

            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockSend();

            const file = await storage.write({ ...metafile, body: readStream, start: 0 });

            expect(file.status).toBe("part");
            expect(file.bytesWritten).toBe(5);
        });

        it("should return bytesWritten when resuming write operation", async () => {
            expect.assertions(1);

            const file = await storage.write({ ...metafile });

            expect(file.bytesWritten).toBe(0);
        });

        it("should reject write operation when file is not found", async () => {
            expect.assertions(1);

            storage.cache.delete(metafile.id);

            // Mock the meta storage get method to simulate file not found
            const mockMetaGet = vi.spyOn(storage.meta, "get");

            mockMetaGet.mockRejectedValueOnce(new Error("File not found"));

            try {
                await storage.write({ ...metafile });

                expect.fail("Expected write to reject");
            } catch (error: any) {
                expect(error).toHaveProperty("UploadErrorCode", "FileNotFound");
            } finally {
                mockMetaGet.mockRestore();
            }
        });

        it.todo("should reject write operation when filesystem errors occur", async () => {
            expect.assertions(1);

            const fileWriteStream = new FileWriteStream();

            vi.spyOn(fs, "createWriteStream").mockImplementationOnce(() => fileWriteStream as unknown as fs.WriteStream);

            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockPipeError(fileWriteStream);

            const write = storage.write({ ...metafile, body: readStream, start: 0 });

            await expect(write).rejects.toHaveProperty("UploadErrorCode", "FileError");
        });

        it.todo("should close file and reset bytesWritten when write is aborted", async () => {
            expect.assertions(2);

            const fileWriteStream = new FileWriteStream();

            vi.spyOn(fs, "createWriteStream").mockImplementationOnce(() => fileWriteStream as unknown as fs.WriteStream);

            const close = vi.spyOn(fileWriteStream, "close");

            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockAbort();

            const file = await storage.write({ ...metafile, body: readStream, start: 0 });

            expect(file.bytesWritten).toBeNaN();
            expect(close).toHaveBeenCalledOnceExactlyOnceWith();
        });

        it.todo("should reject write operation when range is invalid", async () => {
            expect.assertions(1);

            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockSend();

            await expect(() => storage.write({ ...metafile, body: readStream, start: (metafile.size as number) - 2 })).rejects.toThrow("File conflict");
        });

        it("should support file locking to prevent concurrent writes", async () => {
            expect.assertions(1);

            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockSend();

            const write = storage.write({ ...metafile, body: readStream, start: 0 });
            const write2 = storage.write({ ...metafile, body: readStream, start: 0 });

            await expect(Promise.all([write, write2])).rejects.toHaveProperty("UploadErrorCode", "FileLocked");
        });
    });

    describe(".list()", () => {
        beforeEach(() => createFile(false));

        it("should return all user files in storage", async () => {
            expect.assertions(4);

            const items = await storage.list();

            expect(items).toHaveLength(1);
            expect(items[0].id).toBe("/anonymous/testfile.mp4");
            expect(items[0].createdAt).toBeDefined();
            expect(items[0].modifiedAt).toBeDefined();
        });
    });

    describe(".delete()", () => {
        beforeEach(createFile);

        it("should mark file as deleted and return file data", async () => {
            expect.assertions(2);

            const deleted = await storage.delete(metafile);

            expect(deleted.id).toBe(metafile.id);
            expect(deleted.status).toBe("deleted");
        });

        it("should handle deletion of non-existent files gracefully", async () => {
            expect.assertions(2);

            const { promises: fsp } = await import("node:fs");
            const mockReadFile = vi.spyOn(fsp, "readFile");

            mockReadFile.mockRejectedValueOnce("notfound");

            const deleted = await storage.delete({ id: "notfound" });

            expect(deleted.id).toBe("notfound");
            expect(deleted.status).toBeUndefined();
        });
    });

    describe(".purge()", () => {
        beforeEach(() => createFile(false));

        it("should delete expired files based on age threshold", async () => {
            expect.assertions(1);

            vi.useFakeTimers();
            vi.advanceTimersByTime(6000); // Advance by 6 seconds

            const list = await storage.purge(5); // Purge files older than 5 seconds

            expect(list.items).toHaveLength(1);

            vi.useRealTimers();
        });
    });

    describe(".copy()", () => {
        beforeEach(async () => {
            storage = new DiskStorage(options);

            await storage.create(request, deepClone(metafile));
        });

        it("should copy file to new location", async () => {
            expect.assertions(1);

            // Copy the metadata file to a new location
            await storage.copy(`${metafile.id}.META`, `${directory}/newname1.META`);

            // List should still show only 1 item since it only shows data files, not metadata
            const list = await storage.list();

            expect(list).toHaveLength(1);
        });
    });

    describe(".move()", () => {
        beforeEach(async () => {
            storage = new DiskStorage(options);

            await storage.create(request, deepClone(metafile));
        });

        it("should move file to new location", async () => {
            expect.assertions(1);

            await storage.move(`${metafile.id}.META`, `${directory}/newname2.META`);

            const list = await storage.list();

            expect(list).toHaveLength(1);
        });
    });

    describe("streaming functionality", () => {
        describe(".getStream()", () => {
            // TODO Create dummpy mp4 file
            beforeEach(createFile);

            it("should return streaming file data for existing files", async () => {
                expect.assertions(5);

                const result = await storage.getStream({ id: metafile.id });

                expect(result.stream).toBeDefined();
                expect(result.stream.readable).toBe(true);
                expect(result.size).toBeDefined();
                expect(result.headers).toBeDefined();
                expect(result.headers["Content-Type"]).toBe(metafile.contentType);
            });

            it("should return proper headers for streaming", async () => {
                expect.assertions(4);

                const result = await storage.getStream({ id: metafile.id });

                expect(result.headers["Content-Type"]).toBe(metafile.contentType);
                expect(result.headers["Content-Length"]).toBe(metafile.size.toString());
                expect(result.headers["X-Upload-Expires"]).toBeDefined();
                expect(result.headers["Last-Modified"]).toBeUndefined();
            });

            it.todo("should handle files with expiry dates", async () => {
                expect.assertions(1);

                // Update file metadata to include expiry
                const expiredFile = { ...metafile, expiredAt: new Date(Date.now() + 86_400_000) };

                await storage.meta.save(metafile.id, expiredFile);

                const result = await storage.getStream({ id: metafile.id });

                expect(result.headers["X-Upload-Expires"]).toBe(expiredFile.expiredAt.toISOString());
            });

            it.todo("should throw FileNotFound error for non-existent files", async () => {
                expect.assertions(1);

                await expect(storage.getStream({ id: "nonexistent" })).rejects.toThrow("Not found");
            });

            it("should stream large files efficiently", async () => {
                expect.assertions(2);

                // Create a larger file for testing
                const largeContent = Buffer.alloc(1024 * 1024); // 1MB

                fs.writeFileSync(join(directory, `${metafile.id}.bin`), largeContent);

                const largeFile = {
                    ...metafile,
                    name: `${metafile.id}.bin`,
                    size: largeContent.length,
                };

                await storage.meta.save(largeFile.id, largeFile);

                const result = await storage.getStream({ id: largeFile.id });

                expect(result.size).toBe(largeContent.length);
                expect(result.stream).toBeDefined();
            });
        });

        describe("integration with get()", () => {
            beforeEach(createFile);

            it("should return same data via get() and getStream()", async () => {
                expect.assertions(3);

                const bufferResult = await storage.get({ id: metafile.id });
                const streamResult = await storage.getStream({ id: metafile.id });

                expect(streamResult.size).toBe(bufferResult.size);
                expect(streamResult.headers["Content-Type"]).toBe(bufferResult.contentType);

                // Read stream data
                const chunks: Buffer[] = [];

                for await (const chunk of streamResult.stream) {
                    chunks.push(chunk);
                }

                const streamData = Buffer.concat(chunks);

                expect(streamData.equals(bufferResult.content)).toBe(true);
            });
        });
    });
});
