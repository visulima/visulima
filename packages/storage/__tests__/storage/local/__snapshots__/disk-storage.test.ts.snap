// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DiskStorage .create() should reject on limits 1`] = `[ValidationError]`;

exports[`DiskStorage .create() should set status and bytesWritten 1`] = `
File {
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "expiredAt": "2022-02-02T01:00:00.000Z",
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "anonymous/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
}
`;

exports[`DiskStorage .list() should return all user files 1`] = `
[
  {
    "createdAt": 2022-02-02T00:00:00.000Z,
    "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
    "modifiedAt": 2022-02-02T00:00:00.000Z,
  },
]
`;

exports[`disk > .create() > should create disk file with correct status and bytesWritten 1`] = `
File {
  "ETag": undefined,
  "bytesWritten": 0,
  "content": undefined,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "expiredAt": undefined,
  "hash": undefined,
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "modifiedAt": undefined,
  "name": "anonymous/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
}
`;

exports[`disk > .create() > should reject creation when file size exceeds limits 1`] = `[ValidationError]`;

exports[`disk > .list() > should return all user files in storage 1`] = `
[
  {
    "createdAt": 2022-02-02T00:00:00.000Z,
    "id": "/anonymous/testfile.mp4",
    "modifiedAt": 2022-02-02T00:00:00.000Z,
  },
]
`;
